compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -541,6 +538,46 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         }
     }
+
+    /// A captured place is mutable if
+    /// 1. Projections don't include a Deref of an immut-borrow, **and**
+    /// 2. PlaceBase is mut or projections include a Deref of a mut-borrow.
+    fn determine_capture_mutability(&self, place: &Place<'tcx>) -> hir::Mutability {
+        let var_hir_id = match place.base {
+            PlaceBase::Upvar(upvar_id) => upvar_id.var_path.hir_id,
+            _ => unreachable!(),
+        };
+
+        let bm = *self
+            .typeck_results
+            .borrow()
+            .pat_binding_modes()
+            .get(var_hir_id)
+            .expect("missing binding mode");
+
+        let mut is_mutbl = match bm {
+            ty::BindByValue(mutability) => mutability,
+            ty::BindByReference(_) => hir::Mutability::Not,
+        };
+
+        for pointer_ty in place.deref_tys() {
```

Comment:
we should add tests for the following scenarios:

- given `x: &&mut u32`, attempting to mutate `**x += 10` from within a closure should fail
- given `x: &mut &mut u32`, mutating `**x += 10` should work
- given `x: &mut &u32`, mutating `*x = &44` should work
- given `x: &mut &u32`, mutating `**x += 10` should fail